### PR TITLE
202307-notebook-unionall

### PR DESCRIPTION
**Pull request by petermarshallio**
A Jupyter Lab notebook on the UNION ALL operator in Druid, differentiating between top and table level execution and query patterns.

<hr>

This PR has:

- [x] been self-reviewed.
- [x] been tested in a test Druid cluster.
- [ ] been tested in the notebook docker container.

### CODE REVIEW

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1beb4ad7-93cf-4922-9f57-d7b9739fea4a",
+   "metadata": {},
+   "source": [
+    "Optionally, run the next cell – it shows the precise `EXPLAIN PLAN` for the query. You can see there are two `query` exeuction plans, one for each of our queries, and that Druid's planning process has taken time to optimize how the query above will actually execute."
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:226]: Possibly worth having a brief description of what EXPLAIN PLAN means? Instead or in addition, you could mention it's how Druid converts the SQL query into a native one

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 2]
```diff
@@ -0,0 +1,644 @@
...
+   "source": [
+    "print(json.dumps(json.loads(sql_client.explain_sql(sql)['PLAN']), indent=2))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "800add1a-d459-4796-b974-b2f094db417f",
+   "metadata": {},
+   "source": [
+    "The next cell contains two result sets: `set1` provides some statistics for the 1st November, `set2` for the 2nd November.\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:244]: Could you label in the results which are from set1 and which are from set2? It would also help to clarify which set this statement applies to, since one set of results has them in the expected order
>Why are the longest and shortest the wrong way around?!
image

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+    "WITH\n",
+    "set1 AS (\n",
+    "    SELECT \"Reporting_Airline\", \"Distance\"\n",
+    "    FROM \"On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11\"\n",
+    "    WHERE DATE_TRUNC('HOUR', __time) = TIMESTAMP '2005-11-01 11:00:00'\n",
+    "  ),\n",
+    "set2 AS (\n",
+    "    SELECT\n",
+    "        \"Reporting_Airline\",\n",
+    "        COUNT(*) AS \"Frights\",\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:199]: Typo? Fright --> flight

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 4]
```diff
@@ -0,0 +1,644 @@
...
+    "        MAX(Distance) AS \"Lengthiest\",\n",
+    "        MIN(Distance) AS \"Shortest\"\n",
+    "    FROM \"On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11\"\n",
+    "    WHERE DATE_TRUNC('HOUR', __time) = TIMESTAMP '2005-11-02 11:00:00'\n",
+    "    AND \"Reporting_Airline\" LIKE 'A%'\n",
+    "    GROUP BY 1\n",
+    "  )\n",
+    "\n",
+    "SELECT \"Reporting_Airline\",\n",
+    "        COUNT(*) AS \"Frights\",\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:209]: Typo? Fright --> flight

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 5]
```diff
@@ -0,0 +1,644 @@
...
+    "        MAX(Distance) AS \"Longest\"\n",
+    "    FROM \"On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11\"\n",
+    "    WHERE DATE_TRUNC('HOUR', __time) = TIMESTAMP '2005-11-01 11:00:00'\n",
+    "    AND \"Reporting_Airline\" LIKE 'A%'\n",
+    "    GROUP BY 1\n",
+    "  ),\n",
+    "set2 AS (\n",
+    "    SELECT\n",
+    "        \"Reporting_Airline\",\n",
+    "        COUNT(*) AS \"Frights\",\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:272]: Typo? Fright --> flight

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 6]
```diff
@@ -0,0 +1,644 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "f2c95ffc-b260-4671-bacc-c8cc3137e9c2",
+   "metadata": {},
+   "source": [
+    "As expected, we see a concatenated set of results - but wait! Why are the longest and shortest the wrong way around?!\n",
+    "\n",
+    "As the `UNION ALL` concatenated the sets, it very simply added the results for Atlanta to the end of the results for San Francisco. It did not take into account that the columns in set 2 were in a different order, nor did it take into account _either_ of the errors in field names.\n",
+    "\n",
+    "Instead, the query against `set2` ought to have been more explicit, taking into account the proper field names (`flights` vs `frights`!) and giving the same order as the first set of results - ie, `set1`'s schema:\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:298]: I'm not sure I understand this section. Was the typo intentional to describe the mis-labeling with UNION ALL? It's hard to parse what is the wrong label vs. what is the UNION ALL behavior.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 7]
```diff
@@ -0,0 +1,644 @@
...
+    "    \"Flight_Number_Reporting_Airline\"\n",
+    "FROM (\n",
+    "    SELECT * from set1\n",
+    "    UNION ALL\n",
+    "    SELECT * from set2\n",
+    "    )\n",
+    "ORDER BY __time\n",
+    "```\n",
+    "\n",
+    "Run the cell below to see what happens."
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:372]: Preface that this query shouldn't work? Something as subtle as, "However, see what happens when Druid tries to run the following query" or more obvious like "However, Druid does not support this query pattern. See what happens:"

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 8]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "25001794-e1a7-4325-adb3-2b8f26036261",
+   "metadata": {},
+   "source": [
+    "Notice the `ClientError` that Druid returns. Rather than simply concatenating two `TABLE`s of results together, Druid switched to table level `UNION ALL`, and with it a switch to [`UNION`](https://druid.apache.org/docs/26.0.0/querying/datasource.html#union) datasources. These cannot be used in the same way as `TABLE`s; there are constraints on what can be done with them. An `EXPLAIN PLAN` for these types of queries shows this in detail:\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:425]:
```suggestion
    "Notice the `ClientError` that Druid returns. Rather than simply concatenating two `TABLE`s of results together, Druid switched to table level `UNION ALL`. The native query translation uses [`UNION`](https://druid.apache.org/docs/26.0.0/querying/datasource.html#union) datasources. These cannot be used in the same way as `TABLE` datasources; there are constraints on what can be done with them. An `EXPLAIN PLAN` for these types of queries shows this in detail:\n",
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 6 above)
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:298]: Prefer to write out the transition phrase instead of using "ie" or "eg"

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 9]
```diff
@@ -0,0 +1,644 @@
...
+    "    \"type\": \"table\",\n",
+    "    \"name\": \"On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11\"\n",
+    "    },\n",
+    "    {\n",
+    "    \"type\": \"table\",\n",
+    "    \"name\": \"On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11\"\n",
+    "    }\n",
+    "```\n",
+    "\n",
+    "In large deployments of Druid, data engineers may decide to break up a single data set into multiple `TABLE`s. Some reasons for this might be:\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:440]:
```suggestion
    "In large deployments of Druid, data engineers may decide to break up a single data set into multiple `TABLE` datasources. The following examples list some use cases:\n",
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 10]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "* `TABLE`s separated by their `__time` granularity (ie. the level of roll-up)\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:442]: The styling looks odd for `TABLE`s; suggest using `TABLE` datasources instead

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 11]
```diff
@@ -0,0 +1,644 @@
...
+    "* `TABLE`s logically separated for security reasons by tenant\n",
+    "* `TABLE`s broken up by different filter conditions\n",
+    "* `TABLE`s separated by upstream data source (e.g. different Kafka topics for different cloud regions, different groups of source device, and so on)\n",
+    "* `TABLE`s that break the entire data set up into different periods of time, perhaps with different retention periods\n",
+    "\n",
+    "The next two cells to create two new tables, `wikipedia-en` and `wikipedia-fr`. One table contains only English language channels, while the second only contains French language channels. Imagine that this is a design decision taken by a data engineer so that they can be governed separately. But now we want to have a single view of both that people can query.\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:448]:
```suggestion
    "The next two cells create two new tables, `wikipedia-en` and `wikipedia-fr`. One table contains only English language channels, while the second contains only French language channels. Imagine that this is a design decision taken by a data engineer so that they can be governed separately. But now we want to have a single view of both that people can query.\n",
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+    "sql_client.wait_until_ready('wikipedia-fr')\n",
+    "display.table('wikipedia-fr')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f8bbf2c6-681a-46f5-82f2-201cbbe8058d",
+   "metadata": {},
+   "source": [
+    "In the next cell, we declare our `UNION ALL` data source as the single, unified view of all the data. Then we use that in our outer `SELECT` to count the number of robot and non-robot edits by channel.\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:560]:
```suggestion
    "The next cell creates `unifiedSource` using `UNION ALL`. This datasource is the single, unified view of all the data. You then use the unified datasource in a `SELECT` query to count the number of robot and non-robot edits by channel.\n",
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 13]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "Remember, the `SELECT` in the `unifiedSource` must be simple in order to meet the constraints set by a table level `UNION ALL`, so any filtering ans so on can only be done in the outer `SELECT` statement."
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:562]:
```suggestion
    "Remember, the `SELECT` in the `unifiedSource` must be simple in order to meet the constraints set by a table level `UNION ALL`, so any operations such as filtering can only be done in the outer `SELECT` statement."
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 14]
```diff
@@ -0,0 +1,644 @@
...
+    "## Conclusion\n",
+    "\n",
+    "* There are two modes for `UNION ALL` in Druid - top level and table level\n",
+    "* Top level is a simple concatenation, and operations must be done on the source `TABLE`s\n",
+    "* Table level uses a `union` data source, and operations must be done on the outer `SELECT`\n",
+    "\n",
+    "## Learn more\n",
+    "\n",
+    "* Watch [Plan your Druid table datasources](https://youtu.be/OpYDX4RYLV0?list=PLDZysOZKycN7MZvNxQk_6RbwSJqjSrsNR) by Peter Marshall\n",
+    "* Read [union](https://druid.apache.org/docs/26.0.0/querying/datasource.html#union) data sources in the documentation\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:618]:
```suggestion
    "* Read about [union](https://druid.apache.org/docs/26.0.0/querying/datasource.html#union) datasources in the documentation\n",
```

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 15]
```diff
@@ -0,0 +1,644 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "f58a1846-5072-4495-b840-a620de3c0442",
+   "metadata": {},
+   "source": [
+    "## Conclusion\n",
+    "\n",
+    "* There are two modes for `UNION ALL` in Druid - top level and table level\n",
+    "* Top level is a simple concatenation, and operations must be done on the source `TABLE`s\n",
+    "* Table level uses a `union` data source, and operations must be done on the outer `SELECT`\n",
```
Comment (vtlim) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:613]: I really like this conclusion block. I was confused by some of the later examples, so these points weren't very clear to me before this. Conceptually, you can add the following to make it easier to follow:
* Include a table of contents section and more descriptive section headers. Right now we have the following:
   * **Concatenating result sets with UNION ALL** --> consider referring to this as "Top-level"
   * **Working with concatenated result sets** --> consider referring to this as "Table-level"
* Add a bit more details to the introductory sentence. For example, "This notebook introduces the UNION ALL operator and shows two contexts for how this operator can be used: in top-level queries and in table-level queries."
* Add more description to table-level queries. Top-level ones seem to be more intuitive, but it could help to have more details around what table-level queries are, after this sentence (and before jumping into the example)
> Instead of a top level UNION ALL, it switches to a "[table level](https://druid.apache.org/docs/26.0.0/querying/sql.html#table-level)" query.

**Review by vtlim — COMMENTED** [review]
Great start on the tutorial for understanding UNION ALL.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 16]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "If you do not use the Docker Compose environment, you need the following:\n",
+    "\n",
+    "* A running Druid instance.\n",
+    "* [druidapi](https://github.com/apache/druid/blob/master/examples/quickstart/jupyter-notebooks/druidapi/README.md), a Python client for Apache Druid. Follow the instructions in the Install section of the README file.\n",
+    "* [matplotlib](https://matplotlib.org/), a library for creating visualizations in Python,\n",
+    "* [pandas](https://pandas.pydata.org/), a data analysis and manipulation tool.\n",
+    "* Jupyter notebook or Jupyter Lab. See [jupyter.org](https://jupyter.org/) for installation instructions.\n",
+    "\n",
+    "</details>"
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:36]: @sergioferragut looks like the template for notebooks may need to be updated? I will amend this one and the other couple of PRs I have open to remove this for now

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 17]
```diff
@@ -0,0 +1,644 @@
...
+    "### Load example flight data\n",
+    "\n",
+    "Once your Druid environment is up and running, ingest the sample data for this tutorial.\n",
+    "\n",
+    "Open the Druid console:\n",
+    "\n",
+    "1. Load data\n",
+    "2. Batch - SQL\n",
+    "3. Example data\n",
+    "4. Select \"FlightCarrierOnTime (1 month)\"\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:88]: Thanks @vtlim ! I'll make the same amendments to the other notebooks with the flight data loading in it.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 18]
```diff
@@ -0,0 +1,644 @@
...
+    "\n",
+    "`On_Time_Reporting_Carrier_On_Time_Performance_(1987_present)_2005_11`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3b3c7fc0-cb5c-43a0-aa53-2c62053181b0",
+   "metadata": {},
+   "source": [
+    "When this is completed, run the following cell for the final part of the initialization. This will provide us some methods to call as we explore what TopN does."
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:97]: Thanks! I have added a `display.table()` to the notebook. This is possibly a good practice for all notebooks where we rely on some table or other? @sergioferragut @techdocsmith @writer-jill

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 15 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:622]: Thank you! I ended up removing the trap that generates the error condition, and pulling a lot back up under the main headings.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 19]
```diff
@@ -0,0 +1,644 @@
...
+   "id": "f2a19226-6abc-436d-ac3c-9c04d6026707",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import matplotlib\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd"
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb]: Ooooh good catch thank you!

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 3 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:184]: That's intentional.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 4 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:194]: That's intentional.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 2 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:229]: I've updated the text here so that it better highlights what happens and why – and why the following SQL fixes it.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 5 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:257]: Intentional.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 9 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb]: Updated the text in this section.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb (hunk 10 above)
Comment (petermarshallio) [reply examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb]: Updated the text in this section.

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 20]
```diff
@@ -0,0 +1,541 @@
...
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f2c95ffc-b260-4671-bacc-c8cc3137e9c2",
+   "metadata": {},
+   "source": [
+    "Look carefully at these results - results from `set2` were appended very simply.\n",
```
Comment (techdocsmith) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:278]:
```suggestion
    "Look carefully at these results - Druid has simply appended the results from `set2` to `set1` without introducing redundant columns.\n",
```
nit. passive + "very"
Great demo of this behavior!

File: examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb [hunk 21]
```diff
@@ -0,0 +1,541 @@
...
+    "\n",
+    "* Column name in `set2` (`Period` versus `Episode` and `Average Comment Size` versus `Average Comment Length`) did not result in new columns\n",
+    "* Columns with the same name (`Pages`) did not result in that aggregate being put into same column - Austria's values are simply appended `Average Comment Size`\n",
+    "\n",
+    "Run the next cell, which uses explicit column names at the top-level, rather than `*`, to ensure the calculations appear in the right columns in the final result. It also explicitly names the column names for the results by using `AS`."
```
Comment (techdocsmith) [inline examples/quickstart/jupyter-notebooks/notebooks/03-query/04-UnionOperations.ipynb:283]:
```suggestion
    "Run the next cell, which uses explicit column names at the top-level, rather than `*`, to ensure the calculations appear in the right columns in the final result. It also aliases the columns for the results by using `AS`."
```
I think this is called "aliasing", but be sure

**Review by techdocsmith — CHANGES_REQUESTED** [review]

**Review by techdocsmith — CHANGES_REQUESTED** [review]
Awesome work here. This one is ready to merge after comments.
